Output flux diagnostic with checkpoints

roelof-groenewald opened the pull request:
Flux diagnostic data will now be copied to the checkpoint folder after creating a checkpoint. Functionality was also added to delete old checkpoints.

EZoni committed:
Add Python Wrappers for F,G in PML (#2481)

* Add Python Wrappers for F,G in PML

* Add Getters for F,G Nodal Flags

* Fix Bug in <F,G>FPPMLWrapper (Default Level)

* Fix Bug in F,G Nodal Flags

* Use GetPML Method for F,G Nodal Flags

EZoni committed:
PICMI: Add max_grid_size, blocking_factor in (x,y,z) (#2524)

* PICMI: Add amr.max_grid_size_<x,y,z>

* Set All Flags in Python

* PICMI: Add amr.blocking_factor_<x,y,z>

roelof-groenewald committed:
added wrappers to get particle structs for the particles in the boundary buffers (#2498)

ax3l committed:
Doxygen: Fix Docs (#2526)

EZoni committed:
Fix Bug with Tilebox for G in PML (#2527)

* Fix Bug with Tilebox for G in PML

* Reset Benchmark

ax3l committed:
AMReX/PICSAR: Weekly Update (#2533)

Weekly update to latest AMReX.
Weekly update to latest PICSAR (no changes).

```
./Tools/Release/updatePICSAR.py
./Tools/Release/updateAMReX.py
```

ax3l committed:
Install pre-commit (#2532)

* Add pre-commit

Add basis for automated pre-commit checks.
Install locally via:

```bash
python3 -m pip install -U pre-commit
pre-commit install
```

See: https://pre-commit.com

* Cleanup: Whitespaces

* Cleanup: requirements.txt order

ax3l committed:
Summit: Update Numpy Hints (#2535)

Make sure `numpy` can be rebuilt when and were needed.
To achieve that, move numpy-specific installation hints on OpenBLAS
to the WarpX profile.

roelof-groenewald committed:
copy flux diagnostic to checkpoint folders

roelof-groenewald committed:
updated changelog and version number

roelof-groenewald committed:
do not count scraped particles from seed plasma in current calculations

lucafedeli88 committed:
Fix some issues with Fujitsu compiler (#2529)

* make some code compilable with Fujitsu compiler in clang mode

* update documentation

RemiLehe committed:
Fix #2522: Gaussian beam positions do not change with warpx.random_seed (#2523)

* Draw Gaussian beam position with amrex random engine

* Update benchmarks

* Update tolerance in space-charge tests

* Update benchmark for space charge initialization test

* Update benchmarks

* Update benchmark

* Clean-up code

* Update benchmarks

PhilMiller committed:
#2534: Don't access position vector values beyond the configured dimension (#2536)

* #2534: Don't access position vector values beyond the configured dimension

* Fix particle position component used in XZ configuration

* Handle 1D case

* Move values only used in scraping function into inside-boundary condition

* Error out if scraping from EB in RZ

roelof-groenewald committed:
fixed bug in reloading flux history

KZhu-ME commented on mewarpx/mewarpx/assemblies.py:
```diff
@@ -123,6 +123,20 @@ def record_scrapedparticles(self):
+            # show very large currents. For this reason we skip that first step
+            # but note that we inject after the first step so we need to skip
+            # scraping step 2.
+            if data[1] == 2:
```
Are `it=0` and `it=1` skipped elsewhere?

roelof-groenewald replied:
Nope. I think WarpX starts at step 1, and we inject at the end of the step and scrape at the start.

KZhu-ME commented on mewarpx/mewarpx/utils_store/init_restart_util.py:
```diff
@@ -3,38 +3,51 @@
+    """
+    # handle the case where num_to_keep is 0 or None
+    if not num_to_keep:
+        num_to_keep = None
```
Interesting, I didn't know you could slice lists using `None`, and it has the same effect as leaving it blank.

KZhu-ME commented on mewarpx/tests/test_checkpoint_and_restart.py:
```diff
@@ -72,10 +72,52 @@ def test_create_checkpoints():
     run = get_run()
 
-    checkpoint = CheckPointDiagnostic(DIAG_STEPS, CHECKPOINT_NAME)
+    checkpoint = CheckPointDiagnostic(
```
For the sake of test coverage, we should make one of the tests have `clear_old_checkpoints=True`.

KZhu-ME reviewed (commented):
Looks good, just one comment and one suggestion.

roelof-groenewald committed:
implemented code review suggestion from KZ

ax3l committed:
Spack Development: macOS & GNUmake (#2545)

- Add macOS hints for OpenMP
- Add hints for running GNUmake regression tests locally

ax3l committed:
Regression Tests: OMP on (#2548)

We generally run only with one OpenMP threads at the moment, but
disabling OpenMP altogether causes an extra compile, which slows down
CI.

roelof-groenewald committed:
Merge remote-tracking branch 'public/memaster' into rg/flux_reload_fix

roelof-groenewald committed:
Merge remote-tracking branch 'upstream/development' into rg/flux_reload_fix

roelof-groenewald committed:
Merge branch 'memaster' into rg/flux_reload_fix

roelof-groenewald committed:
updated failing two embedded boundary test reference values

KZhu-ME reviewed (approved):
Looks good to me!

peterscherpelz commented on mewarpx/mewarpx/diags_store/checkpoint_diagnostic.py:
```diff
@@ -13,32 +15,91 @@
+        self.flux_diag = None
 
-        super(CheckPointDiagnostic, self).__init__(diag_steps, **kwargs)
+        super(CheckPointDiagnostic, self).__init__(diag_steps+1, **kwargs)
```
Is it possible this should be
```suggestion
        if kwargs.pop('diag_step_offset', 1) != 1:
            raise ValueError(
                "diag_step_offset in CheckPointDiagnostic must be 1.")
        super(CheckPointDiagnostic, self).__init__(
            diag_steps=diag_steps, diag_step_offset=1, **kwargs)
```
Otherwise it seems like this will run 1 step after the checkpoint file, then 2 steps  after the checkpoint file is written, then 3 after etc.

roelof-groenewald replied:
Oh yeah, good point!

peterscherpelz commented on mewarpx/mewarpx/diags_store/checkpoint_diagnostic.py:
```diff
@@ -13,32 +15,91 @@
+
+        # Copy flux diagnostics, if present, to load when restarting.
+        if (self.flux_diag is not None
+            and self.flux_diag.last_run_step == mwxrun.get_it() - 1
```
Wouldn't this line plus the diag_steps+1 period mean this only triggers on the very first dump? Which means maybe your unit test should also cover two diagnostic periods, not 1, since it doesn't catch this.

peterscherpelz replied:
I think if you have MAX_STEPs in the unit test go to 9 rather than 8 you should be able to check for the restart file existing at step 8, if my glance at that file was correct.

roelof-groenewald replied:
I'll mark this as resolved since the fix you pointed out above will take care of this as well.

peterscherpelz reviewed (changes requested):
All looks good except I'm worried about off-by-1 errors as you'll see in my comments.

Also, how would the restart functionality handle a situation where the checkpoint is written but the program ends before the flux is written? Ideally it would fail to restart from that point and use the previous dump instead. Is that how it would work?

roelof-groenewald committed:
Apply suggestions from code review

Co-authored-by: Peter Scherpelz <31747262+peterscherpelz@users.noreply.github.com>

roelof-groenewald commented:
@peterscherpelz To answer your question. Currently, if a failure happens between writing the checkpoint and copying the flux diagnostic, the restart would fail but the program will just exit - it won't automatically restart from the previous checkpoint. We can add logic to automatically delete a checkpoint that it was able to restart from but didn't have flux diagnostics included. Automatically trying to restart from the next checkpoint will take a bit more work. Lets discuss it tomorrow.

peterscherpelz commented:
Sounds good! I'm not positive it's worth it, especially as exiting is the
correct failure mode, but as I said it did actually impact one of my Warp
runs.
One thought I had was to write an empty "checkpoint_complete" file to the
checkpoint directory after the flux diag is written. That would be a
relatively atomic operation. Then it could just check for that file; if it
doesn't exist, don't use that checkpoint and start from the previous.

On Thu, Nov 11, 2021 at 8:40 PM Roelof Groenewald ***@***.***>
wrote:

> @peterscherpelz <https://github.com/peterscherpelz> To answer your
> question. Currently, if a failure happens between writing the checkpoint
> and copying the flux diagnostic, the restart would fail but the program
> will just exit - it won't automatically restart from the previous
> checkpoint. We can add logic to automatically delete a checkpoint that it
> was able to restart from but didn't have flux diagnostics included.
> Automatically trying to restart from the next checkpoint will take a bit
> more work. Lets discuss it tomorrow.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ModernElectron/WarpX/pull/101#issuecomment-966815820>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHSGZPXVWX3E3ZADFUIVFHDULSLDBANCNFSM5HUGJUDA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

roelof-groenewald committed:
fixed issue that caused CI test to fail

roelof-groenewald commented:
@peterscherpelz I guess we could just use the existence of the flux diagnostic file instead of having to write a new `checkpoint_complete` file. To make it atomic I think we can copy it with a different name and then rename once the copy is complete. However, this will mean that we cannot restart runs that don't use flux diagnostics. The reason for this is that in our current setup the simulation has to be initialized (or restarted) before flux diagnostics are initialized. So in other words the simulation doesn't know whether or not flux diagnostics will be included later when it is trying to restart.

peterscherpelz commented:
Yeah, your last point is good, it's always preferable to be able to remove
components without major side effects. I think keeping it like it is, with
an exception if the flux diagnostic is not found or fails to load, makes
sense.

On Fri, Nov 12, 2021 at 9:45 AM Roelof Groenewald ***@***.***>
wrote:

> @peterscherpelz <https://github.com/peterscherpelz> I guess we could just
> use the existence of the flux diagnostic file instead of having to write a
> new checkpoint_complete file. To make it atomic I think we can copy it
> with a different name and then rename once the copy is complete. However,
> this will mean that we cannot restart runs that don't use flux diagnostics.
> The reason for this is that in our current setup the simulation has to be
> initialized (or restarted) before flux diagnostics are initialized. So in
> other words the simulation doesn't know whether or not flux diagnostics
> will be included later when it is trying to restart.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ModernElectron/WarpX/pull/101#issuecomment-967299330>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHSGZPRZVCTBTXFDKW4KGLLULVHCDANCNFSM5HUGJUDA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>
